controllers/rate_limiting_wasmplugin_controller.go

```diff
@@ -0,0 +1,369 @@
+		return nil, err
+	}
+
+	wasmPlugin.Spec.PluginConfig = pluginConfigStruct
+
+	// controller reference
+	if err := r.SetOwnerReference(gw, wasmPlugin); err != nil {
+		return nil, err
+	}
+
+	return wasmPlugin, nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) wasmPluginConfig(ctx context.Context, gw *gatewayapiv1.Gateway) (*wasm.Plugin, error) {
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	wasmPlugin := &wasm.Plugin{
+		FailureMode:       wasm.FailureModeDeny,
+		RateLimitPolicies: make([]wasm.RateLimitPolicy, 0),
+	}
+
+	gatewayAPITopology, err := r.gatewayAPITopologyFromGateway(ctx, gw)
+	if err != nil {
+		return nil, err
+	}
+
+	rateLimitPolicies := gatewayAPITopology.PoliciesFromGateway(gw)
+
+	logger.V(1).Info("wasmPluginConfig", "#RLPS", len(rateLimitPolicies))
+
+	// Shallow copy by assignment
+	rateLimitPoliciesSorted := rateLimitPolicies
+
+	// Sort RLPs for consistent comparison with existing objects
+	sort.Sort(common.PolicyByKey(rateLimitPoliciesSorted))
+
+	for _, policy := range rateLimitPolicies {
+		rlp := policy.(*kuadrantv1beta2.RateLimitPolicy)
+		wasmRLP, err := r.WASMRateLimitPolicy(ctx, gatewayAPITopology, rlp, gw)
+		if err != nil {
+			return nil, err
+		}
+
+		if wasmRLP == nil {
+			// skip this RLP
+			continue
+		}
+
+		wasmPlugin.RateLimitPolicies = append(wasmPlugin.RateLimitPolicies, *wasmRLP)
+	}
+
+	return wasmPlugin, nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) gatewayAPITopologyFromGateway(ctx context.Context, gw *gatewayapiv1.Gateway) (*common.KuadrantTopology, error) {
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	routeList := &gatewayapiv1.HTTPRouteList{}
+	// Get all the routes having the gateway as parent
+	err = r.Client().List(ctx, routeList, client.MatchingFields{HTTPRouteParents: client.ObjectKeyFromObject(gw).String()})
+	logger.V(1).Info("gatewayAPITopologyFromGateway: list httproutes from gateway", "err", err)
+	if err != nil {
+		return nil, err
+	}
+
+	rlpList := &kuadrantv1beta2.RateLimitPolicyList{}
+	// Get all the rate limit policies
+	// TODO(eastizle): Add index field??
+	err = r.Client().List(ctx, rlpList)
+	logger.V(1).Info("gatewayAPITopologyFromGateway: list rate limit policies", "err", err)
+	if err != nil {
+		return nil, err
+	}
+
+	return common.NewKuadrantTopology(
+		[]*gatewayapiv1.Gateway{gw},
+		common.Map(routeList.Items, func(r gatewayapiv1.HTTPRoute) *gatewayapiv1.HTTPRoute { return &r }),
+		common.Map(rlpList.Items, func(p kuadrantv1beta2.RateLimitPolicy) common.KuadrantPolicy { return &p }),
+	), nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) WASMRateLimitPolicy(ctx context.Context, t *common.KuadrantTopology, rlp *kuadrantv1beta2.RateLimitPolicy, gw *gatewayapiv1.Gateway) (*wasm.RateLimitPolicy, error) {
+	gwHostnamesTmp := common.TargetHostnames(gw)
+	gwHostnames := common.Map(gwHostnamesTmp, func(str string) gatewayapiv1.Hostname { return gatewayapiv1.Hostname(str) })
```

Comment:
This seems to be the reverse of https://github.com/Kuadrant/kuadrant-operator/blob/d4758585cfe4f17efd74207042991722ce06bc09/pkg/common/common.go#L187

Reply:
Yeah, I am aware there is a little mess around the type of the hostnames values. Some functions/APIs require `string` while other require `gatewayapiv1.Hostname`. 

In this particular case, The gateway holds `gatewayapiv1.Hostname`, but the method `TargetHostnames` converts them to `string`. The wasm plugin wants `string` but before that we need to process them and filter valid subdomains and the method ` FilterValidSubdomains` requires `gatewayapiv1.Hostname`. 

Too many back and forth conversions and copies. Unacceptable for a processing in the data plane. I guess not a big issue in the control plane.